lib/lte_uart_dfu/lte_uart_dfu.c

```diff
@@ -0,0 +1,340 @@
+	context.download_started = false;
+	context.download_success = fnc->data.success;
+
+	return 0;
+}
+
+static void execute_input_fnc(const struct DFU_UART_target_RemoteFnc *fnc)
+{
+	int err;
+	size_t offs_get;
+	static size_t progress;
+
+	LOG_DBG("DFU function called: %d", fnc->type_choice);
+
+	switch (fnc->type_choice) {
+	case _DFU_UART_target_RemoteFnc_type_init:
+		err = lte_uart_dfu_parse_init(fnc);
+		if (err) {
+			LOG_ERR("Could not parse init, error %d", err);
+			lte_uart_dfu_send_error(
+				_DFU_UART_target_RemoteFnc_type_init, err);
+			break;
+		}
+
+		progress = 0;
+		err = dfu_target_mcuboot_init(context.file_size,
+					      context.image_num,
+					      unused_func);
+		lte_uart_dfu_send_error(_DFU_UART_target_RemoteFnc_type_init,
+					err);
+		break;
+	case _DFU_UART_target_RemoteFnc_type_write:
+		err = lte_uart_dfu_parse_write(fnc);
+		if (err) {
+			LOG_ERR("lte_uart_dfu_parse_write returned %d", err);
+			lte_uart_dfu_send_error(
+				_DFU_UART_target_RemoteFnc_type_write, err);
+			break;
+		}
+
+		err = dfu_target_mcuboot_write(context.to_write,
+					       context.to_write_len);
+		progress += context.to_write_len;
+		LOG_INF("Progress: %d %%", progress * 100 / context.file_size);
+		lte_uart_dfu_send_error(_DFU_UART_target_RemoteFnc_type_write,
+					err);
+		break;
+	case _DFU_UART_target_RemoteFnc_type_offset:
+		err = dfu_target_mcuboot_offset_get(&offs_get);
+		lte_uart_dfu_send_offset(err, offs_get);
+		break;
+	case _DFU_UART_target_RemoteFnc_type_done:
+		err = lte_uart_dfu_parse_done(fnc);
+		if (err) {
+			lte_uart_dfu_send_error(
+				_DFU_UART_target_RemoteFnc_type_done, err);
+			break;
+		}
+
+		err = dfu_target_mcuboot_done(context.download_success);
+		if (err) {
+			lte_uart_dfu_send_error(
+				_DFU_UART_target_RemoteFnc_type_done, err);
+			break;
+		}
+
+		err = dfu_target_mcuboot_schedule_update(
+			context.image_num);
+		lte_uart_dfu_send_error(_DFU_UART_target_RemoteFnc_type_done,
+					err);
+		break;
+	default:
+		break;
+	}
+}
+
+static void process_input_data(void)
+{
+	int err;
+	uint8_t *cursor;
+	size_t cursor_len;
+	size_t dec_len = 0;
+	struct DFU_UART_target_RemoteFnc fnc = {0};
+
+	while (true) {
+		cursor_len = ring_buf_get_claim(&context.rx_rb, &cursor, UINT32_MAX);
+		if (cursor_len == 0) {
+			/* Let's reset it to prevent wrapping */
+			ring_buf_reset(&context.rx_rb);
+			return;
+		}
+
+		/* We try to put the whole buffer in the decode function.
+		 * If there is not enough data, the decode will throw an error.
+		 * If there is too much data, no error will be thrown and
+		 * the correct amount of decoded data will be given by the
+		 * decode function
+		 */
+		err = cbor_decode_DFU_UART_target_RemoteFnc(
+			cursor,
+			cursor_len,
+			&fnc,
+			&dec_len);
+		if (err) {
+			ring_buf_get_finish(&context.rx_rb, 0);
+			return;
+		}
+
+		execute_input_fnc(&fnc);
+
+		/* We need to remove the amount of consumed data */
+		err = ring_buf_get_finish(&context.rx_rb, dec_len);
```

Comment:
Shouldn't processing break here? If this fails here, what are the chances that line 260 will be successful?
As far as I can tell, this may lead to assertion getting trigger at line 260, if assertions are enabled.